README.md
# Livewire Smart Table
An advanced, dynamic datatable component with pagination, sorting, and searching including json data.

![Livewire Smart Table Demo](demo/livewire-smart-table.gif)

## Installation

You can install the package via composer:

```bash
composer require tkaratug/livewire-smart-table
```

## Requirements
This package uses livewire/livewire (https://laravel-livewire.com/) under the hood.

It also uses Tailwind (https://tailwindcss.com) for base styling.

Please make sure you include both of these dependencies before using this component.

## Usage
In order to use this component, you must create a new Livewire component that extends from `LivewireSmartTable`

You can use `make:livewire` to create a new component. For example.

```
php artisan make:livewire UserList
```

In the `UserList` class, instead of extending from the base Livewire Component class, extend from `LivewireSmartTable` class.
Also, remove the render method. You'll have a class similar to this snippet. 

In this class, you must define columns that you want to show in a table.

```php
class UserList extends LivewireSmartTable
{
    $columns = [
        'id' => [
            'type' => 'string', // column type
            'name' => 'Id', // column header
            'class' => 'text-danger', // column class
        ],
        'name' => [
            'type' => 'string',
            'name' => 'Name',
        ],
        'email' => [
            'type' => 'string',
            'name' => 'E-Mail',
        ],
    ];
}
```

Keys of columns array must be the same as column names in database table or key of a json object.

To render the component in a view, just use the Livewire tag or include syntax.

```blade
<livewire:user-list :query="$query" />
```

`$query` must be instance of an **Eloquent Collection**.

For example, create a `UserController` class, select users to show in a table and pass them to a view file.

````php
class UserController extends Controller
{
    public function index()
    {
        $users = App\User::where('is_active', '=', true)->get();

        return view('users', ['users' => $users]);
    }
}
````

Then in `users.blade.php` use Livewire tag and give users to `query` attribute.
```
<livewire:user-list :query="$users" />
```

## Column Properties
### ```string```
It is used for showing data as string in HTML table.

### ```link```
It is used for showing data as link in HTML table. 

In addition to type, you must define a `url` to redirect when clicked.
```php
$columns = [
    'profile' => [
        'type' => 'link',
        'url' => 'http://example.com/users/{id}/profile',
        'target' => '_blank'
    ],
];
```

It is also possible to give parameters to the URL. All you need to do is give the column name containing the data you want to pass to the url in curly braces.

Let's say you have a database table contains blog posts and each post has a slug. To show post titles in html table as a link, you need to define column as follows:
```php
'title' => [
    'type' => 'link',
    'url' => 'http://example.com/posts/{slug}',
];
```
The component is smart enough to find the `slug` field of current record and give it to the url.

### ```json```
It is used for showing data from json columns. If you have a json column in your database table, you can show values from it in html table.

Let's say you have a json column named `contact` in your database table and contains address details in it.

`{"address":{"country":"Turkey","city":"Istanbul","state":"Besiktas"}}`

To show just the city in html table, you need to define column as follows:
```php
'city' => [
    'type' => 'json',
    'name' => 'City', // Text for column header
    'from' => 'contact', // field that contains json data in a db table
    'value' => 'address.city' // nested json value
];
```
It will find the json data from `contact` column, and take city value inside address key then show it on the table.

### ```actions```
It is used for showing action links for each row in html table. 

You need to give `element` and `url` keys for the html element of the link and url to redirect.

```php
'actions' => [
    'type' => 'actions',
    'name' => 'Actions', // Text for column header
    'actions' => [
        [
            'element' => '<button>View</button>',
            'url' => 'http://example.com/users/{id}/profile'
        ],
        [
            'element' => '<button>Edit</button>',
            'url' => 'http://example.com/users/{id}/edit'
        ],
    ]
];
```

## Publishing Views
You can also publish the view files to customize them.

All you need to do is running the following command. Then the views will be copied into `/resources/views/vendor/livewire-smart-table` directory.

```bash
php artisan vendor:publish --tag=livewire-smart-table-views
```

## Testing

```bash
composer test
```

## Security

If you discover any security related issues, please email tkaratug@hotmail.com.tr instead of using the issue tracker.

## Credits

- [Turan Karatuğ](https://github.com/tkaratug)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `livewire-smart-table` will be documented in this file

## 1.4.0 - 2022-02-25
- Laravel 9.x support added.
- PHP 7.4, 8.0 and 8.1 support added.
- PHP 7.1 support removed.

## 1.3.0 - 2021-04-10
- Laravel 8.x support added.
- Livewire 2.x support added.
- Styles changed to Tailwind.
- Added support for array casted json fields.

## 1.2.1 - 2020-07-31
- Command for publishing the views has been added to the README.md file.

## 1.2.0 - 2020-06-28
- Tests added

## 1.1.0 - 2020-06-21

- Added column properties.

## 1.0.0 - 2020-06-14

- initial release


tests/TestCase.php
<?php

namespace Tkaratug\LivewireSmartTable\Tests;

use CreateDummyTable;
use Livewire\LivewireServiceProvider;
use Orchestra\Testbench\TestCase as BaseCase;
use Tkaratug\LivewireSmartTable\LivewireSmartTableServiceProvider;

class TestCase extends BaseCase
{
    public function setUp(): void
    {
        parent::setUp();

        include_once __DIR__ . '/database/migrations/create_dummy_table.php';

        (new CreateDummyTable())->up();
    }

    protected function getPackageProviders($application)
    {
        return [
            LivewireServiceProvider::class,
            LivewireSmartTableServiceProvider::class,
        ];
    }

    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('app.key', 'base64:Hupx3yAySikrM2/edkZQNQHslgDWYfiBfCuSThJ5SK8=');
        $app['config']->set('database.default', 'testbench');
        $app['config']->set('database.connections.testbench', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => ''
        ]);
    }
}


tests/database/models/Dummy.php
<?php

namespace Tkaratug\LivewireSmartTable\Tests\Database\Models;

use Illuminate\Database\Eloquent\Model;

class Dummy extends Model
{
    protected $table = 'dummies';

    protected $fillable = ['firstname', 'lastname', 'email'];
}


tests/LivewireSmartTableTest.php
<?php

namespace Tkaratug\LivewireSmartTable\Tests;

use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Foundation\Testing\WithFaker;
use Livewire\Livewire;
use Livewire\LivewireManager;
use Livewire\Testing\TestableLivewire;
use Tkaratug\LivewireSmartTable\LivewireSmartTable;
use Tkaratug\LivewireSmartTable\Tests\Database\Models\Dummy;

class LivewireSmartTableTest extends TestCase
{
    use RefreshDatabase, WithFaker;

    private function createComponent($params): TestableLivewire
    {
        return app(LivewireManager::class)->test(LivewireSmartTable::class, $params);
    }

    /** @test */
    public function can_create_component_with_required_params()
    {
        // Arrange
        $name = $this->faker->name;

        // Act
        $component = $this->createComponent([
            'name' => $name,
        ]);

        // Assert
        $this->assertNotNull($component);

        $component->assertSet('columns', []);
        $component->assertSet('search', '');
        $component->assertSet('page', 1);
        $component->assertSet('perPage', 10);
        $component->assertSet('sortField', 'id');
        $component->assertSet('sortAsc', true);
        $component->assertSet('tableClass', 'table');
        $component->assertSet('sortIcon', '&#8597;');
        $component->assertSet('sortAscIcon', '&#8593;');
        $component->assertSet('sortDescIcon', '&#8595;');
    }

    /** @test */
    public function can_get_data_and_set_columns()
    {
        $data = $this->prepareData();
        $columns = $this->prepareColumns();

        Livewire::test(LivewireSmartTable::class, ['query' => $data])
            ->set('columns', $columns)
            ->assertSet('columns', $columns)
            ->assertSee($data->first()->firstname)
            ->set('page', 2)
            ->assertSee($data->last()->firstname);
    }

    /**
     * Prepare data to list in table.
     *
     * @return \Illuminate\Support\Collection
     */
    private function prepareData()
    {
        for ($i = 1; $i <= 20; $i++) {
            Dummy::create([
                'firstname' => $this->faker->firstName,
                'lastname' => $this->faker->lastName,
                'email' => $this->faker->email
            ]);
        }

        return Dummy::query()->get();
    }

    /**
     * Prepare columns.
     *
     * @return array
     */
    private function prepareColumns()
    {
        return [
            'id' => [
                'type' => 'string',
                'name' => 'id',
            ],
            'firstname' => [
                'type' => 'string',
                'name' => 'First Name',
            ],
        ];
    }
}


src/LivewireSmartTableServiceProvider.php
<?php

namespace Tkaratug\LivewireSmartTable;

use Illuminate\Support\ServiceProvider;

class LivewireSmartTableServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        $this->loadViewsFrom(__DIR__.'/../resources/views', 'livewire-smart-table');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('livewire-smart-table.php'),
            ], 'config');

            // Publishing the views.
            $this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/livewire-smart-table'),
            ], 'livewire-smart-table-views');
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'livewire-smart-table');
    }
}


resources/views/livewire-smart-table.blade.php
<div>
    @if ($data->isEmpty())
        <div class="bg-red-100 border border-red-200 text-red-600 p-3 rounded relative" role="alert">
            <span class="block sm:inline">No data have found.</span>
        </div>
    @else
        <div class="container mx-auto py-6 px-4">
            <div class="mb-4 flex justify-between items-center">
                <div class="flex-1 pr-4">
                    <div class="relative md:w-1/3">
                        <input
                            wire:model.debounce:500ms="search"
                            type="search"
                            class="w-full pl-10 pr-4 py-2 rounded-lg shadow focus:outline-none focus:shadow-outline text-gray-600 font-medium"
                            placeholder="Search...">
                        <div class="absolute top-0 left-0 inline-flex items-center p-2">
                            <svg xmlns="http://www.w3.org/2000/svg" class="w-6 h-6 text-gray-400" viewBox="0 0 24 24"
                                 stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round"
                                 stroke-linejoin="round">
                                <rect x="0" y="0" width="24" height="24" stroke="none"></rect>
                                <circle cx="10" cy="10" r="7" />
                                <line x1="21" y1="21" x2="15" y2="15" />
                            </svg>
                        </div>
                    </div>
                </div>

                <div>
                    <div class="shadow rounded-lg flex">
                        <div class="relative">
                            <div class="relative">
                                <select wire:model="perPage" class="appearance-none border-none inline-block py-2 pl-4 pr-4 rounded leading-tight w-full">
                                    <option>10</option>
                                    <option>20</option>
                                    <option>30</option>
                                </select>
                            </div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="bg-white rounded-lg shadow Ûrelative">
                <table class="border-collapse table-auto w-full whitespace-no-wrap bg-white table-striped relative {{ $tableClass }}">
                    <thead>
                    <tr class="text-left">
                        @foreach ($columns as $column => $props)
                            <th class="bg-gray-100 sticky top-0 border-b border-gray-200 px-6 py-2 text-gray-600 font-bold tracking-wider uppercase text-xs">
                                @if ($props['type'] !== 'actions')
                                    <a wire:click.prevent="sortBy('{{ $column }}')" role="button" href="#">
                                        @if (isset($props['name']))
                                            {{ $props['name'] }}
                                        @else
                                            {{ $column }}
                                        @endif
                                    </a>
                                    @include('livewire-smart-table::sort-icon', ['field' => $column])
                                @else
                                    @if (isset($props['name']))
                                        {{ $props['name'] }}
                                    @else
                                        {{ $column }}
                                    @endif
                                @endif
                            </th>
                        @endforeach
                    </tr>
                    </thead>
                    <tbody>
                    @foreach ($data as $item)
                        <tr>
                            @foreach ($columns as $column => $props)
                                @if ($props['type'] === 'link')
                                    <td class="border-dashed border-t border-gray-200 {{ $props['class'] ?? '' }}">
                                        <a href="{{ $item->{$column.'_url'} }}" target="{{ $props['target'] ?? '_blank' }}">
                                            {!! $item->{$column} !!}
                                        </a>
                                    </td>
                                @elseif ($props['type'] === 'actions')
                                    <td class="border-dashed border-t border-gray-200 {{ $props['class'] ?? '' }}">
                                        @foreach ($item->{$column} as $option)
                                            <a href="{{ $option['url'] }}">{!! $option['element'] !!}</a>
                                        @endforeach
                                    </td>
                                @else
                                    <td class="border-dashed border-t border-gray-200 {{ $props['class'] ?? '' }}">
                                        <span class="text-gray-700 px-6 py-3 flex items-center">{!! $item->{$column} !!}</span>
                                    </td>
                                @endif
                            @endforeach
                        </tr>
                    @endforeach
                    </tbody>
                </table>
            </div>

            <div class="mt-4">
                {{ $data->links() }}
            </div>
        </div>
    @endif
</div>


src/LivewireSmartTable.php
<?php

namespace Tkaratug\LivewireSmartTable;

use Illuminate\Contracts\View\Factory;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Pagination\LengthAwarePaginator;
use Illuminate\View\View;
use Livewire\Component;
use Livewire\WithPagination;

class LivewireSmartTable extends Component
{
    use WithPagination;

    public Collection $query;

    public string $search = '';

    public $page = 1;

    public int $perPage = 10;

    public string $sortField = 'id';

    public bool $sortAsc = true;

    public array $columns = [];

    public string $tableClass = 'table';

    public string $sortIcon = '&#8597;';

    public string $sortAscIcon = '&#8593;';

    public string $sortDescIcon = '&#8595;';

    public function mount(Collection $query)
    {
        $this->query = $query;
        $this->page = request()->query('page', $this->page);
    }

    /**
     * Sort by.
     *
     * @param $field
     * @return void
     */
    public function sortBy($field): void
    {
        if ($this->sortField === $field) {
            $this->sortAsc = ! $this->sortAsc;
        }

        $this->sortField = $field;
    }

    /**
     * Render
     *
     * @return Factory|View
     */
    public function render()
    {
        $items = $this->prepareData($this->query);

        // If user search something, filter items by input
        if (!empty($this->search)) {
            $items = $this->query->filter(function ($item) {
                $found = false;
                foreach ($this->columns as $column => $props) {
                    if ($props['type'] !== 'actions') {
                        if (stripos($item->{$column}, $this->search) !== false) {
                            $found = true;
                        }
                    }
                }
                return $found;
            });
        }

        // Sort results
        $data = $this->sortAsc
            ? $items->sortBy($this->sortField)
            : $items->sortByDesc($this->sortField);

        // Paginate results
        $data = new LengthAwarePaginator(
            $data->forPage($this->page, $this->perPage),
            $data->count(),
            $this->perPage,
            $this->page,
            ['path' => request()->url(), 'query' => request()->query()]
        );

        return view('livewire-smart-table::livewire-smart-table', [
            'data'    => $data,
            'columns' => $this->columns,
        ]);
    }

    /**
     * Prepare data by columns
     *
     * @param Collection $query
     * @return Collection
     */
    private function prepareData(Collection $query): Collection
    {
        foreach ($query as $item) {
            foreach ($this->columns as $key => $props) {
                if (isset($props['type'])) {
                    switch ($props['type']) {
                        case 'json':
                            $from = $item->hasCast($props['from'], ['array'])
                                ? $item->{$props['from']}
                                : json_decode($item->{$props['from']}, true);
                            $nestedFields = explode('.', $props['value']);
                            foreach ($nestedFields as $field) {
                                $from = $from[$field] ?? null;
                            }
                            $item->{$key} = $from;
                            break;
                        case 'link':
                            $item->{$key.'_url'} = $this->makeUrl($props['url'], $item);
                            break;
                        case 'actions':
                            $actions = [];
                            foreach ($props['actions'] as $action) {
                                $actions[] = [
                                    'element' => $action['element'],
                                    'url' => $this->makeUrl($action['url'], $item),
                                ];
                                $item->{$key} = $actions;
                            }
                            break;
                        default:
                    }
                }
            }
        }
        return $query;
    }

    /**
     * Make url
     *
     * @param string $url
     * @param $item
     * @return string
     */
    private function makeUrl(string $url, $item): string
    {
        preg_match_all('/\{.*?\}/', $url, $matches);

        foreach ($matches[0] as $match) {
            preg_match('/\{([^\]]*)\}/', $match, $field);
            $url = str_replace($match, $item->{$field[1]}, $url);
        }

        return $url;
    }
}
